### DISCUSSION TITLE
Rule to prevent parallel pending promises before await

### DISCUSSION THREAD

**Discussion by ackvf** [eslint/eslint#17824] [Ideas] [answered]
In a situation like this, when a promise is awaited, while another fails, it results in an uncaught error.
In the example below, the failed promise is at the line `const r2 = await p2`, the following code never runs and therefore the remaining promises are not awaited.

More explanation https://stackoverflow.com/a/69644993/985454

```js
try {
  const p1 = Promise.resolve('p1:success')
  const p2 = Promise.reject('p2:failure')
  const p3 = new Promise((resolve, _) => {
    setTimeout(() => {
      console.log('p3 finished')
      resolve("p3:success")
    }, 3000)
  })
  const p4 = new Promise((_, reject) => {
    setTimeout(() => {
      console.log('p4 finished')
      reject("p4:failure")
    }, 1000)
  })
  

  const r1 = await p1
  const r2 = await p2 // ❌following code never runs!
  const r3 = await p3
  const r4 = await p4
  console.log('await:', [r1, r2, r3, r4])


  // this would also throw ❌ - ✅ caught by no-floating-promises rule
  // Promise.all([p1, p2, p3, p4]).then((values) => {
  //   console.log('all:', values)
  // })


  // this would not throw ✅
  // await Promise.all([p1, p2, p3, p4]).then((values) => {
  //   console.log('all:', values)
  // })


} catch (e) {
  console.log('catch:', e)
}
```

- My suggestion is that when an `await` is encountered AND there are more than one pending promises, the rule will suggest to use `await Promise.allSettled` or to append `.catch()`

Strangely enough, the `no-floating-promises` rule does not catch these cases
```
const p2 = Promise.reject('p2:failure') // ❌ no error given

const p4 = new Promise((_, reject) => { // ❌ no error given
  setTimeout(() => {
    console.log('p4 finished')
    reject("p4:failure")
  }, 1000)
})

Promise.all([p1, p2, p3, p4]).then((values) => { // ✅ error: Promises must be awaited, end with a call to .catch, ...
  console.log('all:', values)
})
```

**Comment by nzakas** [comment, marked as answer]
You'll need to talk to the typescript-eslint folks about your concerns, as this rule is not part of the ESLint core: https://typescript-eslint.io/rules/no-floating-promises/

**Reply by ackvf** [reply]
~https://github.com/typescript-eslint/typescript-eslint/discussions/8097~
https://github.com/typescript-eslint/typescript-eslint/issues/8098
